### DISCUSSION TITLE
Pl_DCT

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#5] [General]
https://github.com/qpdf/qpdf/pull/1229 is much better than my https://github.com/qpdf/qpdf/pull/1225 . That said, I think this is still not quite right.

Taking a step back, I am not sure that there is a point of trying to decompress damaged JPEGs during fuzzing, either as part of dct_fuzzer or qpdf_fuzzer. Once jpeglib has detected that a JPEG is damaged, calling further jpeglib methods fuzzes jpeglib rather than qpdf. We already know that in this situation jpeglib may be slow and require huge amounts of memory. This suggests to me that there is no point to limit the size of uncompressed corrupt data during fuzzing - throwing an exception once corruption is detected would be the better solution. (There will still be the need to limit memory allocated by jpeglib during fuzzing in order to avoid spurious positives).

The option of throwing an exception when corrupt JPEGs are encountered is probably also appropriate for normal processing by qpdf, as this would cause qpdf to fall back preserve the unfiltered data and continue normally, rather than writing out an uncompressed and potentially huge corrupt JPEG which could make the whole pdf unusable. 

In terms of how to implement the configuration option, what stumped me originally and let me down the path of conditional compilation was that creation of Pl_DCT objects is buried pretty deep in the library without a clean path to configure it from the QPDF level. Looking at 1229 triggered the thought of an alternative solution. We can use a static configuration method to configure at the class rather than instance level, which would also allow qpdf_fuzzer to use this config. I'll do a PR(https://github.com/qpdf/qpdf/pull/1230)  for you to look at.

**Comment by jberkenbilt** [comment]
Sounds good. The thought of having a class/static method crossed my mind as well. Unless I messed up, errors from the jpeg library should already be getting converted to exceptions. I don't remember for sure, but I'm pretty sure there's a test for that.

**Reply by m-holger** [reply]
Errors are already converted, but warnings ("Warnings: the library can continue, but the data is corrupt, and a  damaged output image is likely to result.") are currently ignored, both in the sense that qpdf tries to soldier on and that they are not passed on to the qpdf user (neither as warning nor as return code).

**Comment by jberkenbilt** [comment]
Okay, yes, I suspect it would be best if we would treat those as errors and leave the images unfiltered as we do for errors. I wasn't aware of that issue. Thanks!

**Comment by m-holger** [comment]
I pushed PR https://github.com/qpdf/qpdf/pull/1230
